### PR TITLE
Improvements Try class

### LINKED ISSUE #59: Improvements Try class
- Refactor fun names and order
- Refactor tests (one test per fun) and improve coverage
- Add mapError() that should merge exceptions using CompositeException

### PR DESCRIPTION
**Pull request by Ts1n**
- Refactor fun names and order
- Refactor tests (one test per fun) and improve coverage
- Add mapError() that should merge exceptions using CompositeException

#closes #59

### CODE REVIEW

File: src/main/kotlin/org/ionproject/integration/utils/Try.kt [hunk 1]
```diff
@@ -15,13 +15,18 @@ private fun merge(e1: Exception, e2: Exception) =
     }
 
 sealed class Try<out T> {
+    abstract fun get(): T
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/utils/Try.kt]: Isn't this the same as orThrow in terms of usage?

File: src/main/kotlin/org/ionproject/integration/utils/Try.kt [hunk 2]
```diff
@@ -32,26 +37,29 @@ sealed class Try<out T> {
...
-        fun <T> of(value: T) = Value(value)
-        fun ofError(e: Exception) = Error(e)
         fun <T> of(f: () -> T): Try<T> = try {
             Value(f())
         } catch (e: Exception) {
             Error(e)
         }
 
+        fun <T> ofValue(value: T) = Value(value)
+        fun <T> ofError(e: Exception) = Error<Exception>(e)
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/utils/Try.kt]: Could this be ... = Error< T >(e), to make sure the Try is specifically typed to match the type of e?

File: src/main/kotlin/org/ionproject/integration/utils/Try.kt [hunk 3]
```diff
@@ -32,26 +37,29 @@ sealed class Try<out T> {
         override fun <R> match(success: (T) -> R, error: (Exception) -> R) = success(value)
     }
 
-    class Error(val e: Exception) : Try<Nothing>() {
+    class Error<out T>(val e: Exception) : Try<T>() {
```
Comment (sscosta) [inline src/main/kotlin/org/ionproject/integration/utils/Try.kt]: Couldn't T be bound to be an exception, like so: class Error < out T: Exception> (val e : Exception) ?

File: src/test/kotlin/org/ionproject/integration/utils/TryTests.kt [hunk 4]
```diff
@@ -2,28 +2,146 @@ package org.ionproject.integration.utils
...
+
+    @Test
+    fun whenDivideByZeroAndMapError_thenReturnsErrorWithCompositeMessage() {
+        // Act
+        val result = oper(1, 0)
+            .mapError { NullPointerException("NullPointer Exception") }
+            .match({ it.toString() }, { it.javaClass.simpleName })
+
+        // Assert
+        assertEquals("CompositeException", result)
```
Comment (sscosta) [inline src/test/kotlin/org/ionproject/integration/utils/TryTests.kt]: In this case we know what are the types of the exceptions enclosed: NPE and ArithmeticException. Do you think we could test it to make sure the contract is not broken?

**Review by sscosta — DISMISSED** [review]

File: src/main/kotlin/org/ionproject/integration/utils/Try.kt (hunk 3 above)
Comment (Ts1n) [reply src/main/kotlin/org/ionproject/integration/utils/Try.kt]: No because then R would also need to be an exception (because it's the return of map, flatmap,..) and that would break Value

**Review by sscosta — DISMISSED** [review]

File: src/main/kotlin/org/ionproject/integration/utils/Try.kt [hunk 5]
```diff
@@ -32,26 +35,28 @@ sealed class Try<out T> {
...
```
Comment (pmhsfelix) [inline src/main/kotlin/org/ionproject/integration/utils/Try.kt]: Why this change?

Comment (Ts1n) [reply src/main/kotlin/org/ionproject/integration/utils/Try.kt]: I've made the change while trying to implement mapError and forgot to change it back.

**Review by sscosta — APPROVED** [review]
LGTM

**Review by pmhsfelix — APPROVED** [review]
